library/minecraft_server_file.py

```diff
@@ -0,0 +1,354 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+"""
+Generate Minecraft ACLs.
+"""
+
+import grp
+import json
+import os
+import pwd
+import stat
+import tempfile
+import uuid
+
+from datetime import datetime as dt
+
+MINECRAFT_API_URL = 'https://api.mojang.com/profiles/minecraft'
+SERVER_FILE_CHOICES = ['ops', 'whitelist', 'banned-players', 'banned-ips',
+                       'server-properties']
+DEFAULT_BAN_EXPIRES = 'forever'
+DEFAULT_BAN_REASON = 'Banned by an operator.'
+MINECRAFT_OP_CODE = 4
+
+
+class MissingArgsException(Exception):
+
+    def __init__(self, missing_args):
+        super(MissingArgsException, self).__init__(
+            message='missing required arguments: {}'.format(
+                ', '.join(missing_args)))
+
+
+class BadValueException(Exception):
+
+    def __init__(self, bad_value):
+        super(BadValueException, self).__init__(
+            message='bad value: {}'.format(bad_value))
+
+
+class FileStats(object):
+    """
+    Compare and update file stats.
+    """
+    REQUIRED_ARGS = ['owner', 'group']
+
+    def __init__(self, module):
+        self.module = module
+        missing_args = [x for x in self.REQUIRED_ARGS
+                        if not self.module.params[x]]
+        if missing_args:
+            raise MissingArgsException(missing_args)
+
+        self.path = os.path.expanduser(module.params['path'])
+        self.owner = module.params['owner']
+        self.group = module.params['group']
+        self.mode = module.params['mode'] if module.params['mode'] else 0644
+
+    def file_stats_changed(self):
+        try:
+            st = os.stat(self.path)
+        except OSError:
+            # If the file doesn't exist or isn't read, this is the same as
+            # the stat values being wrong.
+            return True
+
+        current_mode = st.st_mode & stat.S_IMODE(st.st_mode)
+        # We cannot look up non-existant uids and gids when we compare the
+        # current state to desired state. Instead, get the usernames and
+        # groups now to make comparison easier.
+        current_uid, current_gid = self.module.user_and_group(self.path)
+        current_user = pwd.getpwuid(current_uid).pw_name
+        current_group = grp.getgrgid(current_gid).gr_name
+
+        return any([
+            self.owner != current_user,
+            self.group != current_group,
+            self.mode != current_mode,
+        ])
+
+    def correct_file_stats(self):
+        self.module.set_owner_if_different(self.path, self.owner, False)
+        self.module.set_group_if_different(self.path, self.group, False)
+        self.module.set_mode_if_different(self.path, self.mode, False)
+
+
+class ServerFile(object):
```

Comment:
It didn't seem necessary; the abc module doesn't appear to buy us anything here. From my limited experience, abc is mostly concerned with inspection of objects, which we don't do. I can add it if you like, but it doesn't seem like it'd buy us much, right now.